core/src/main/scala/org/http4s/internal/Macros.scala

```diff
@@ -10,11 +10,47 @@ private[http4s] object Macros {
     s
   }
 
-  def scheme(c: Context)(): c.Expr[Scheme] = {
+  def fragment(c: Context)(): c.Expr[Uri.Fragment] = {
     import c.universe._
     val s = literalString(c)
-    Scheme.parse(s) match {
-      case Right(_) => c.Expr(q"""org.http4s.HttpCodec[org.http4s.Scheme].parseOrThrow($s)""")
+    Uri.Fragment.parse(s) match {
+      case Right(_) => c.Expr(q"""org.http4s.HttpCodec[org.http4s.Uri.Fragment].parseOrThrow($s)""")
+      case Left(e) => throw e
+    }
+  }
+
+  def host(c: Context)(): c.Expr[Uri.Host] = {
```

Comment:
@ChristopherDavenport brings up that these can interpolate.  I think that's true, for some types, if we can determine that the surrounding context is valid.  That would make these macros super powerful, but it's going to require a lot of careful thought.

An example for this one: an arbitrary string is a `RegHost`, but we can't allow it if the host starts with `[` or `]`, which denotes an IPv6 or IPvFuture address.